circuits/cpp/src/aztec3/circuits/kernel/public/common.hpp

```diff
@@ -0,0 +1,213 @@
+    auto& preimages = public_kernel_inputs.public_call.public_call_data.public_call_stack_preimages;
+    for (size_t i = 0; i < stack.size(); ++i) {
+        const auto& hash = stack[i];
+        const auto& preimage = preimages[i];
+
+        // Note: this assumes it's computationally infeasible to have `0` as a valid call_stack_item_hash.
+        // Assumes `hash == 0` means "this stack item is empty".
+        const auto calculated_hash = hash == 0 ? 0 : preimage.hash();
+        composer.do_assert(hash != calculated_hash,
+                           format("public_call_stack[", i, "] = ", hash, "; does not reconcile"));
+    }
+};
+
+template <typename NT, template <class> typename KernelInput>
+void validate_function_execution(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates state reads and transitions for all type of kernel inputs
+    validate_state_reads(composer, public_kernel_inputs);
+    validate_state_transitions(composer, public_kernel_inputs);
+}
+
+template <typename NT>
+void validate_this_public_call_hash(DummyComposer& composer, PublicKernelInputs<NT> const& public_kernel_inputs)
+{
+    // Pops the current function execution from the stack and validates it against the call stack item
+    const auto& start = public_kernel_inputs.previous_kernel.public_inputs.end;
+    // TODO: this logic might need to change to accommodate the weird edge 3 initial txs (the 'main' tx, the 'fee' tx,
+    // and the 'gas rebate' tx).
+    const auto popped_public_call_hash = array_pop(start.public_call_stack);
+    const auto calculated_this_public_call_hash =
+        public_kernel_inputs.public_call.public_call_data.call_stack_item.hash();
+
+    composer.do_assert(
+        popped_public_call_hash == calculated_this_public_call_hash,
+        "calculated public_call_hash does not match provided public_call_hash at the top of the call stack");
+};
+
+template <typename NT, template <class> typename KernelInput>
+void common_validate_kernel_execution(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates kernel execution for all type of kernel inputs
+    validate_this_public_call_stack(composer, public_kernel_inputs);
+
+    validate_function_execution(composer, public_kernel_inputs);
+}
+
+template <typename NT, template <class> typename KernelInput>
+void common_validate_inputs(DummyComposer& composer, KernelInput<NT> const& public_kernel_inputs)
+{
+    // Validates commons inputs for all type of kernel inputs
+    const auto& this_call_stack_item = public_kernel_inputs.public_call.public_call_data.call_stack_item;
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_delegate_call == false,
+                       "Users cannot make a delegatecall");
+    composer.do_assert(this_call_stack_item.public_inputs.call_context.is_static_call == false,
+                       "Users cannot make a static call");
```

Comment:
These two checks (delegate call and static call) may be `true` for all cases except if in the "no previous kernel" case. (The "no previous kernel" case means a user is making the call), in which case this check is correct.
We can't do delegate calls or static calls from Noir Contracts, yet, so if this is too tricky to refactor as part of this PR, we should make an issue to come back to it.

Reply:
Not tricky at all. I've localised the validations to the case with no previous kernel.